Source/Core/VideoBackends/D3D12/D3DBase.cpp

```diff
@@ -0,0 +1,859 @@
+		D3D12_ROOT_SIGNATURE_FLAG_DENY_HULL_SHADER_ROOT_ACCESS;
+
+	textRootSignatureDesc.NumParameters = ARRAYSIZE(rootParameters);
+
+	if (!g_ActiveConfig.bEnablePixelLighting)
+		textRootSignatureDesc.NumParameters--;
+
+	ID3DBlob* textRootSignatureBlob;
+	ID3DBlob* textRootSignatureErrorBlob;
+
+	CheckHR(PD3D12SerializeRootSignature(&textRootSignatureDesc, D3D_ROOT_SIGNATURE_VERSION_1, &textRootSignatureBlob, &textRootSignatureErrorBlob));
+
+	CheckHR(D3D::device12->CreateRootSignature(0, textRootSignatureBlob->GetBufferPointer(), textRootSignatureBlob->GetBufferSize(), IID_PPV_ARGS(&defaultRootSignature)));
+}
+
+void WaitForOutstandingRenderingToComplete()
+{
+	commandListMgr->ClearQueueAndWaitForCompletionOfInflightWork();
+}
+
+void Close()
+{
+	// we can't release the swapchain while in fullscreen.
+	swapchain->SetFullscreenState(false, nullptr);
+
+	// Release all back buffer references
+	for (UINT i = 0; i < ARRAYSIZE(backbuf); i++)
+	{
+		SAFE_RELEASE(backbuf[i]);
+	}
+
+	commandListMgr->ImmediatelyDestroyAllResourcesScheduledForDestruction();
+
+	SAFE_RELEASE(swapchain);
+
+	commandListMgr->Release();
+	commandQueue->Release();
+
+	defaultRootSignature->Release();
+
+	gpuDescriptorHeapMgr->Release();
+	samplerDescriptorHeapMgr->Release();
+	rtvDescriptorHeapMgr->Release();
+	dsvDescriptorHeapMgr->Release();
+
+	D3D::CleanupPersistentD3DTextureResources();
+
+	ULONG references12 = device12->Release();
+	if ((!debug12 && references12) || (debug12 && references12 > 1))
+	{
+		ERROR_LOG(VIDEO, "Unreleased D3D12 references: %i.", references12);
+	}
+	else
+	{
+		NOTICE_LOG(VIDEO, "Successfully released all D3D12 device references!");
+	}
+
+#if defined(_DEBUG) || defined(DEBUGFAST)
+	if (debug12)
+	{
+		--references12; // the debug interface increases the refcount of the device, subtract that.
+		if (references12)
+		{
+			// print out alive objects, but only if we actually have pending references
+			// note this will also print out internal live objects to the debug console
+			debug12->ReportLiveDeviceObjects(D3D12_RLDO_DETAIL);
+		}
+		SAFE_RELEASE(debug12);
+	}
+#endif
+
+	device12 = nullptr;
+	currentCommandList = nullptr;
+
+	// unload DLLs
+	UnloadDXGI();
+	UnloadD3DCompiler();
+	UnloadD3D();
+}
+
+const char* VertexShaderVersionString()
+{
+	if (featlevel == D3D_FEATURE_LEVEL_11_0) return "vs_5_0";
+	else if (featlevel == D3D_FEATURE_LEVEL_10_1) return "vs_4_1";
+	else /*if(featlevel == D3D_FEATURE_LEVEL_10_0)*/ return "vs_4_0";
+}
+
+const char* GeometryShaderVersionString()
+{
+	if (featlevel == D3D_FEATURE_LEVEL_11_0) return "gs_5_0";
+	else if (featlevel == D3D_FEATURE_LEVEL_10_1) return "gs_4_1";
+	else /*if(featlevel == D3D_FEATURE_LEVEL_10_0)*/ return "gs_4_0";
+}
+
+const char* PixelShaderVersionString()
+{
+	if (featlevel == D3D_FEATURE_LEVEL_11_0) return "ps_5_0";
+	else if (featlevel == D3D_FEATURE_LEVEL_10_1) return "ps_4_1";
+	else /*if(featlevel == D3D_FEATURE_LEVEL_10_0)*/ return "ps_4_0";
```

Comment:
The bodies should all be on their own line, same as above for the other two functions